Add logging of Spark execution plan to facilitate debugging

DJSagarAhire opened the pull request:
This PR adds logging for the execution plan for Spark pipelines. The logging can be seen in the full raw logs.

Example log:
```
2022-08-11 20:59:59,771 - DEBUG [spark-submitter-phase-1-725c5b7c-198a-11ed-8838-000000bbd65a:i.c.c.e.s.b.RDDUtils@82] - (200) MapPartitionsRDD[49] at flatMapToPair at BaseRDDCollection.java:237 []
  |   MapPartitionsRDD[48] at flatMap at BaseRDDCollection.java:162 []
  |   MapPartitionsRDD[47] at map at BaseRDDCollection.java:157 []
  |   MapPartitionsRDD[46] at map at RDDCollection.java:289 []
  |   MapPartitionsRDD[45] at map at RDDCollection.java:288 []
  |   MapPartitionsRDD[44] at javaRDD at RDDCollection.java:287 []
  |   MapPartitionsRDD[43] at javaRDD at RDDCollection.java:287 []
  |   MapPartitionsRDD[42] at javaRDD at RDDCollection.java:287 []
  |   ZippedPartitionsRDD2[41] at javaRDD at RDDCollection.java:287 []
  |   MapPartitionsRDD[35] at javaRDD at RDDCollection.java:287 []
  |   ShuffledRowRDD[34] at javaRDD at RDDCollection.java:287 []
  +-(200) MapPartitionsRDD[33] at javaRDD at RDDCollection.java:287 []
      |   MapPartitionsRDD[32] at javaRDD at RDDCollection.java:287 []
      |   MapPartitionsRDD[28] at createDataFrame at RDDCollection.java:397 []
      |   MapPartitionsRDD[27] at map at RDDCollection.java:396 []
      |   MapPartitionsRDD[26] at map at RDDCollection.java:113 []
      |   MapPartitionsRDD[25] at flatMap at BaseRDDCollection.java:162 []
      |   MapPartitionsRDD[24] at toJavaRDD at RDDCollection.java:477 []
      |   MapPartitionsRDD[23] at toJavaRDD at RDDCollection.java:477 []
      |   MapPartitionsRDD[22] at toJavaRDD at RDDCollection.java:477 []
      |   MapPartitionsRDD[21] at toJavaRDD at RDDCollection.java:477 []
      |   MapPartitionsRDD[20] at toJavaRDD at RDDCollection.java:477 []
      |   ShuffledRowRDD[19] at toJavaRDD at RDDCollection.java:477 []
      +-(20) MapPartitionsRDD[16] at toJavaRDD at RDDCollection.java:477 []
         |   MapPartitionsRDD[15] at toJavaRDD at RDDCollection.java:477 []
         |   MapPartitionsRDD[14] at toJavaRDD at RDDCollection.java:477 []
         |   MapPartitionsRDD[13] at toJavaRDD at RDDCollection.java:477 []
         |   MapPartitionsRDD[12] at toJavaRDD at RDDCollection.java:477 []
         |   MapPartitionsRDD[11] at toJavaRDD at RDDCollection.java:477 []
         |   MapPartitionsRDD[10] at toJavaRDD at RDDCollection.java:477 []
         |   UnionRDD[9] at union at BaseRDDCollection.java:137 []
         |   MapPartitionsRDD[2] at flatMap at BaseRDDCollection.java:162 []
         |   MapPartitionsRDD[1] at flatMap at BatchSparkPipelineDriver.java:128 []
         |   DatasetRDD[0] at RDD at DatasetRDD.scala:52 []
         |   MapPartitionsRDD[8] at flatMap at BaseRDDCollection.java:162 []
         |   MapPartitionsRDD[7] at flatMap at BatchSparkPipelineDriver.java:128 []
         |   DatasetRDD[6] at RDD at DatasetRDD.scala:52 []
  |   MapPartitionsRDD[40] at javaRDD at RDDCollection.java:287 []
  |   ShuffledRowRDD[39] at javaRDD at RDDCollection.java:287 []
  +-(10) MapPartitionsRDD[37] at javaRDD at RDDCollection.java:287 []
     |   MapPartitionsRDD[36] at javaRDD at RDDCollection.java:287 []
     |   MapPartitionsRDD[31] at createDataFrame at RDDCollection.java:397 []
     |   MapPartitionsRDD[30] at map at RDDCollection.java:396 []
     |   MapPartitionsRDD[29] at map at RDDCollection.java:144 []
     |   MapPartitionsRDD[5] at flatMap at BaseRDDCollection.java:162 []
     |   MapPartitionsRDD[4] at flatMap at BatchSparkPipelineDriver.java:128 []
     |   DatasetRDD[3] at RDD at DatasetRDD.scala:52 []
```

chtyim committed:
Refactoring

- Refactor to cleanup resources at the end of task execution
- Refactor logs in DataprocRuntimeJobManager for files upload
- Rename and move the ProgramRunDispatcherInfo to ProgramRunDispatcherContext
- Fix ProgramRunDispatcherContext to not to serialize Runnable

dli357 committed:
[CDAP-19448] Add interface method for directly retrieving a program classloader from the ProgramRuntimeProvider"

[CDAP-19529] Provide interface for loading TwillRunnable classes to master environment runnable via context

[CDAP-19529] Allow spark core classes to be exposed via SPARK_PROGRAM_CLASS_LOADER_FILTER

[CDAP-19529] Address comments

JessieYee committed:
Introduce MetricRetrievalException and Throw Exception in SparkProgramStatusMetricsProvide with unit test.

vsethi09 committed:
CDAP-19437 Exclude recent log folders during cleanup in LogCleaner service and fix flaky test

tivv reviewed (commented):
I am wondering if we can also put stage names into it by calling org.apache.spark.SparkContext#setCallSite. If would be 100x more valuable

tivv commented:
Probably call setCallSite in the for loop of io.cdap.cdap.etl.spark.SparkPipelineRunner#runPipeline and clear in the end

snehalakshmisha committed:
Enhance EndPoint class to also include location value

de-lan committed:
chore(CDAP-19493): turn on the feature flag for Composite Trigger

de-lan committed:
chore(CDAP-19495): Add feature flag for LCM edit

itsankit-google committed:
try to fetch gpg key from multiple servers

rmstar committed:
[CDAP-19051] Add dataproc operation ID to failure log messages

DJSagarAhire commented:
> I am wondering if we can also put stage names into it by calling org.apache.spark.SparkContext#setCallSite. If would be 100x more valuable

Suggested changes have been added. Logs are now of the format:

```
2022-08-16 17:27:21,048 - DEBUG [spark-submitter-phase-1-90062f64-1d5a-11ed-82d6-00000035c3e1:i.c.c.e.s.b.RDDUtils@83] - (200) MapPartitionsRDD[49] at  []
  |   MapPartitionsRDD[48] at Joiner (Plugin Joiner of Type batchjoiner) []
```

tivv commented on cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/SparkPipelineRunner.java:
```diff
@@ -336,6 +343,7 @@ public void runPipeline(PhaseSpec phaseSpec, String sourcePluginType,
       }
 
       emittedRecords.put(stageName, emittedBuilder.build());
+      jsc.setCallSite("");
```
Just call `clearCallSite`. Also please do it in `try..finally` to ensure it's always cleaned

tivv reviewed (commented):
It looks great, just a minor comment.

tivv commented on cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/SparkPipelineRunner.java:
```diff
@@ -187,9 +191,12 @@ public void runPipeline(PhaseSpec phaseSpec, String sourcePluginType,
 
       EmittedRecords.Builder emittedBuilder = EmittedRecords.builder();
 
+      jsc.setCallSite(stageName + " (Plugin " + pluginName + " of Type " + pluginType + ")");
```
nit: consider adding plugin artifact information (name/version)

tivv replied:
You can even include full `stageSpec.getPlugin().getArtifact().toString()`

rmstar committed:
[CDAP-19474] Ensure that tethering connections are in sync on client and server side

Fixed a bug where the list of namespaces and description go out of sync when tethering is deleted and recreated on the client side with a
different description and namespaces.

de-lan committed:
feat(CDAP-19426): Support adding new columns and indices in StructuredTableAdmin

snehalakshmisha committed:
Add SPI definition for emitting metadata

vsethi09 committed:
[CDAP-19316] Fix security vulnerabilities in commons-collections

vsethi09 committed:
[CDAP-19316] Fix security vulnerabilities in jackson-databind

floratgg0521 committed:
Add sparkmetrics retry policy and host check

itsankit-google committed:
exclude log4j

sanjanasandeep committed:
Squashing commits

Deployment logic for edit flow

Cleaning tests

Cleaning tests

Cleaning more tests and modifying GET /appdetails endpoint

Changes to other Post/PUT apis for deployment

adding new fields to table and updating the get latest function

Refactor AppDeploymentInfo class

- Use builder instead of multiple constructors for optional fields
- Combine the applicationClass and applicationClassName into one field

Removing changes to apis

Making Long non-nullable and removing old addApplication

Addressing comments

addressing 2nd round of comments

addressing 3rd round of comments

Addressing comments and using scan-by-index

CDAP-19465 Refactoring the Kill() method in RemoteExecutionTwillController.complete()

- Adding kill(RuntimeJobDetail) method to TetheringRuntimeJobManager to handle the case
where the Program is completed and therefore a terminate should not be issues.

- Deprecating unused list() method

[CDAP-19459] Remove ProgramRunnerClassLoaderFactory

addressing comments and replace scan with san-by-index

Revert "[CDAP-19459] Remove ProgramRunnerClassLoaderFactory"

This reverts commit 62f34c215ed87d74a252265c89c97d10138a3da9.

Revert "CDAP-19465 Refactoring the Kill() method in RemoteExecutionTwillController.complete()"

This reverts commit 88a25672eda599e9765c2a47191216198d8eaaab.

addressing nitpicks

tivv reviewed (commented):
I don't see changes comparing to the last revision. Did you change anything?

ultizet committed:
Make OAuth API more generic.
This change makes the OAuth API compatible with more systems than just Salesforce (e.g. Hubspot).

ultizet committed:
Change the way the auth URL is formed.

ultizet committed:
Fix the checks for missing JSON fields.

ultizet committed:
Add backward compatibility for the redirect_url parameter.

ultizet committed:
Fix the tests.

floratgg0521 committed:
Enforce at least once correct initialization for SystemMetricsExporter

itsankit-google committed:
exclude log4j

ultizet committed:
Remove the commons-lang dependency.

adinilfeld committed:
Add support for resampling to Wrangler

Co-authored-by: Ellen Liao <elliao@google.com>

jingtaowu3 committed:
CDAP-19540: add unit test for overwriting old tags and properties in create mutation

masoud-io committed:
Compute hashes for artfacts.jar and artifact_archive and caches these artifacts in GCS

de-lan committed:
chore(CDAP-19521): remove the table registry from postgresql admin

albertshau committed:
CDAP-19565 fix backwards incompatible change to storage SPI

Added back the create() method to StructuredTableAdmin and made
the new createOrUpdate() method have a default implementation.
This ensures that the changes to the interface are backwards
compatible.

rmstar committed:
[CDAP-19476] Add logs in TetheringAgentService, TetheringServerHandler

Add logging when tethering control messages are sent and received. This will help identifing cause of delay in program state getting updated.

kapadiap committed:
Adds pipeline runtime argument to prevent cleanup of k8s resources.

This gives a way to keep the pipeline run resources (like pods,
Job, StatefulSet etc.) for debugging failures.

The resources will eventually be deleted when the next
background cleanup thread executes. The background cleanup job's
interval is configurable.

Users will be able to configure individual pipeline run to skip cleanup
of the resources using the property - "system.job.skipCleanup":"true"

This runtime argument is only applicable to jobs running natively
in kubernetes.

Testing Done:
1. Verified with system worker enabled
   - ensure system worker mode is enabled in cdapmasters CR
   - update cdap image in the cluster and restart all relevant pods
   - setup UI and run sample pipeline
   - verify system worker pod logs show pipeline triggered
   - verify 2 pods in the system after job completes -
     ```
     cdap-cdap-workflow-default-datafusionquickstart-2-datapipek7j7h   0/1     Completed   0          14m
     phase-1-726a1a8332c95085-driver                                   0/1     Completed   0          12m
     ```
   - verify the Job resource exists in the cluster.
   - wait for background job to execute and verify pods are gone.
   - app fabric pod logs confirm background job removed all resources.

2. Verified with system worker disabled
   - repeat above steps with system worker mode disabled in cdapmasters CR.
   - verify the app-fabric pod triggers the pipeline run from the logs.

3. update cdapmasters CR spec.configs field to add new config -
   program.container.cleaner.interval.mins:                            420
   restart app-fabric pod and re-run the pipeline. check the resources stay
   longer than the default 60 min.

greeshmaswaminathan committed:
Auto detect ateast once processing mode for streaming pipelines

albertshau committed:
CDAP-19573 increase default resources for test pipelines

Increased the default memory for pipeline configs created
programmatically through the ETLConfig.Builder class. This will
not affect the default value for pipelines created through the
REST API, only the default value for pipelines created
programmatically for tests. This also has no impact on unit tests,
only on integration tests.

de-lan committed:
feat(CDAP-19426): add parentversion in draft store

katukota-google committed:
CDAP-19556 | State store changes in platform

greeshmaswaminathan committed:
Streaming pipeline runner changes for state handling

katukota-google committed:
CDAP-19556 | State store changes in platform

dli357 committed:
[CDAP-19581] Start deployment watcher thread for non-system namespaces and add logs to CDAP resource creation failure in k8s

[CDAP-19581] Add KubeAPIException which wraps ApiException to make the statue code and response body visible in the error message

sahusanket committed:
CDAP-18436 : Added platform changes to support direct reads using the SQL engine : BQ Source Pushdown

itsankit-google committed:
CDAP-19569: cache twill and launcher jar to prevent building for every run

rmstar committed:
[CDAP-19609] Create an access token generator tool

katukota-google committed:
CDAP-19556 | State store changes in platform

jingtaowu3 committed:
CDAP-19536 hide version in MatadataEntity json response

dj-smart committed:
CDAP-19579: Added delegatingOutputStream to handle switches b/w streams
and delegatingLogEvent to propogate mdc tags.

katukota-google committed:
CDAP-19556 | State store changes in platform

tivv committed:
CDAP-19614 - Don't pass ApplicationDescription through the runtime service

tivv committed:
CDAP-19618 Deduplicate strings in MetricValues deserialization

chtyim committed:
Refactor AppDeploymentInfo class

- Use builder instead of multiple constructors for optional fields
- Combine the applicationClass and applicationClassName into one field

katukota-google committed:
CDAP-19556 | State store changes in platform

katukota-google committed:
CDAP-19556 | State store changes in platform

masoud-io committed:
Add hidden Dataproc config to explicitly disable GCS caching for a profile

ameya0111 committed:
CDAP-19465 Refactoring the Kill() method in RemoteExecutionTwillController.complete()

- Adding kill(RuntimeJobDetail) method to TetheringRuntimeJobManager to handle the case
where the Program is completed and therefore a terminate should not be issues.

- Deprecating unused list() method

sahusanket committed:
CDAP-19601 : changing default drop down option for machine type to E2

snehalakshmisha committed:
Add context for Metadata Consumer

snehalakshmisha committed:
Add subscriber service for metadata consumer extensions

kapadiap committed:
Update the deprecated spark configuration key

Addressing the following warning in the logs: `2021-10-13 02:14:35,968 - WARN  [main:o.a.s.SparkConf@69] - The configuration key 'spark.maxRemoteBlockSizeFetchToMem' has been deprecated as of Spark 3.0 and may be removed in the future. Please use the new key 'spark.network.maxRemoteBlockSizeFetchToMem' instead.`

de-lan committed:
feat(CDAP-19624): extend scan and sort by index in StructuredTable - sql/spanner implementation

sanjanasandeep committed:
Adding interface for scan by keyrange and index

masoud-io committed:
Check if delete lifecycle is enabled on bucket before caching artifacts

greeshmaswaminathan committed:
Enable state handling in StreamingContext

sanjanasandeep committed:
NoSql scan by filter and range

janakigg committed:
Introduce ErrorCodeProvider and use it in a few places. CDAP-19162

sanjanasandeep committed:
Adding sort functionality and tests

janakigg committed:
Possible fix for flaky HangingWorkerTest ITN. Increasing stop wait timeout

janakigg committed:
Log ProgramId of App timing out during Stop

katukota-google committed:
CDAP-19562 | Capture metrics for multiple sources, types for atleast once processing used by pipelines

katukota-google committed:
CDAP-19563 | Add warning message as untested usage for more than one source to discourage users

albertshau committed:
CDAP-18510 only skip k8s job cleanup if the job failed

masoud-io committed:
Read before setting custom time to avoid being throttled by GCS

sanjanasandeep committed:
addressed comments

kapadiap committed:
[CDAP-18966] Run kubeJobCleaner only in appfabric pod

Currently the kubeJobCleaner background thread runs in both AppFabric and Preview services where only 1 is required.

Testing Done:

Change cdapmasters CR with a. trace log level for io.cdap.cdap.k8s.runtime. b. reduce cleanup interval with program.container.cleaner.interval.mins: 10
Run pipeline with runtime param system.runtime.cleanup.disabled=true in k8s native env.
observe the appFabric and preview pod logs and confirm cleanup is only running in appfabric.
verify the pipeline specific entities are cleaned up after jobcleaner executes.

chtyim committed:
Collect logs from the twill AM container as part of program execution logs

- Use the EventHander to initialize the CDAP specific log appender to send logs back to CDAP from AM

Highlight of the changes

1. Changing to use the DistributedProgramContainerModule to standardize the guice injection such that it is the same as any other processes we run in Dataproc / Hadoop.
2. Since the Twill library doesn't support file localization to the AM container, I workaround it by using the resources.jar, we have already been using it to localize cConf and hConf XML files to the AM.
3. Since ProgramOptions is a complex object, it is better to just Gson it to a file and deserialize it back from the EventHandler. This is the same mechanism as being used elsewhere in CDAP when running in distributed mode. The serialized file is localized using the resources.jar as mentioned above. Having the program options give us the benefit that most of the program execution related information is there (runId, clusterMode, logger settings... etc).
4. We also need to localize the logback.xml so that logger configuration is consistent with other containers. Again, we use the resources.jar to do so.
5. Since the logback.xml is inside the resources.jar directory, a normal logback library discovery process won't find it. So we have to use the logback.configurationFile property to point it to the correct file path in the AM container directory. However, we need to override it back for other containers (aka TwillRunnables), hence you can see in the code that the mentioned property has been set twice to a different value. 6. The default logback.xml in CDAP, which is used by CDF as well, has a property CDAP_LOG_DIR that we need to set to the YARN <LOG_DIR> so that YARN can substitute it with the container log directory at run time. You can see in the change that I used the command property for that.

One last point: We duplicated the program option json, cConf, hConf, and logback.xml for the TwillRunnable container (one inside the resources.jar, one through normal file localization that lands them in the YARN container directory). We actually can change them to use the ones inside the resources.jar. However, the change would not be so small because the KubeTwillRunnerService implementation doesn't generate the resources.jar (because there is no AM). Since those files are small and to keep the changes minimal, I decided to just let it be.

albertshau committed:
CDAP-19463 make k8s client connect and read timeouts configurable

Added properties for the kubernetes client connect and read
timeouts. Also defaulted the timeouts to much larger numbers to
have a more stable out of the box experience.

sanjanasandeep committed:
addressed new comments

mrahanjam committed:
update locators for tethering e2e tests

rmstar committed:
Fixed a bug in AccessTokenGeneratorService

Fixed a bug where the service terminated immediately after starting.

sanjanasandeep committed:
addressing more comments

dli357 committed:
[CDAP-19459] Refactor ArtifactClassLoaderFactory to use ProgramRunnerClassLoaderFactory

greeshmaswaminathan committed:
Add state store enabled info in streaming context

sanjanasandeep committed:
Addressing 3rd round of comments

sanjanasandeep committed:
addressing 4th round of comments

rmstar committed:
Adjust tethering agent connection interval

Poll every second to reduce latency of program state updates.

kapadiap committed:
Adds null check for podKillerTask

greeshmaswaminathan committed:
Add stage name prefix to the state key for uniqueness in the app

kapadiap committed:
Remove null check for Master Environment Task

dli357 committed:
[CDAP-19459] Remove ProgramRunnerClassLoaderFactory

janakigg committed:
Capture error message in pipeline logs when getClusterStatus detects cluster creation failure CDAP-19606

arjan-bal committed:
Classify dataproc cluster creation and job submission user errors and add links to dataproc troubleshooting docs

dli357 committed:
[CDAP-19659] Refactor SparkResourceFilter and use Spark classloader to instantiate SparkResourceFilter

[CDAP-19659] Add a unit test for testing the scenario in which the parent environment does not have access to Spark classes

[CDAP-19659] Styling fixes and address comments

snehalakshmisha committed:
Update metadata consumer context

udurbha committed:
Windows Aggregation Definition

masoud-io committed:
Add program.jar to gcs caching

albertshau committed:
CDAP-19468 configure Spark to return after submitting the job

Configure Spark to avoid waiting for the job to complete in
SparkSubmit and rely on our own pod watcher to determine if the
job is complete. Since we already are watching the pod, this
reduces the load on Kubernetes. It also bypasses a Spark bug
where it can wait forever even after the job is complete.

Commit:
Garbage Collection of Cached Program Artifacts

sanjanasandeep committed:
changing creationTimeMillis datatype

Commit:
Enabling Artifact Localizer service in system pods

udurbha committed:
Added additional properties and nullable for optional properties such as order fields list and fields which can be set if frame type is row or range for Windows Aggregation Definition

Commit:
Reusing existing blob info while updating temporary hold on GCS object

tivv committed:
CDAP-19676 Fix returning proper next token for NoSQL StructuredTable store (e.g. Sandbox)

itsankit-google committed:
CDAP-19579: upgrade cdap-twil to 1.2.0-SNAPSHOT

itsankit-google committed:
Logback-classic version 1.1.4 and later require slf4j-api version 1.7.15 or later

sanjanasandeep committed:
Add latest field to ChangeDetail in the response

sanjanasandeep committed:
Adding change-detail to GET /apps response

sanjanasandeep committed:
addressed comments

sanjanasandeep committed:
adding latest

sanjanasandeep committed:
Fixing failing test

masoud-io committed:
Fix BooleanUtils exception

DJSagarAhire committed:
Add logging of Spark execution plan: Fix merge conflicts

DJSagarAhire committed:
Add null check on artifact
